### PR TITLE
Initial sample-source example

### PR DESCRIPTION
**Pull request by lberk**
Building off our sample-controller example/template.  This sample
source is a self contained example using a heartbeat style receive
adapter to import the event into the system (using the controller to
create the RA deployment) and providing example.yaml to properly
forward to a ksvc via cloudevent.

OWNERS approvers match working group leads
(I've added myself as a reviewer if appropriate)

squashed into a single commit due to the differing git histories, but
the work was built on https://github.com/openshift-knative/sample-source

### CODE REVIEW

**Comment by matzew** [comment]
> (I've added myself as a reviewer if appropriate)

+1

**Comment by matzew** [comment]
good work on getting this started

/assign @lionelvillard 
/assign @n3wscott 

Let's see that we get this in, as a first step :-)

File: OWNERS [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# The OWNERS file is used by prow to automatically merge approved PRs.
+
+approvers:
+- mattmoor
+- n3wscott
+- vaikas-google
```
Comment (lionelvillard) [inline OWNERS]: remove vaikas-google? @vaikas

Comment (lberk) [reply OWNERS]: I thought about it, but there were some repo's that still have `vaikas-google` so I opted to play it safe (knative/serving for example).  Happy to remove if requested.

**Comment by lionelvillard** [comment]
Someone has to setup prow on this repo, not sure how it works. I wonder if it's possible to run at least the unit tests before merging this PR.

@n3wscott @vaikas @nachocano  do you know?

File: OWNERS (hunk 1 above)
Comment (vaikas) [reply OWNERS]: It's fine to remove, thanks for doing this!

Comment (lberk) [reply OWNERS]: done in 36c9430

**Comment by nachocano** [comment]
> Someone has to setup prow on this repo, not sure how it works. I wonder if it's possible to run at least the unit tests before merging this PR.
> 
> @n3wscott @vaikas @nachocano do you know?

@chizhg can you provide some guidance on setting up PROW in this new repo?

**Comment by lberk** [comment]
prow config PR submitted: https://github.com/knative/test-infra/pull/1562

File: pkg/adapter/adapter.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+func (a *Adapter) newEvent() cloudevents.Event {
+
+	e := cloudevents.Event{
+		Context: cloudevents.EventContextV1{
+			ID:              uuid.New().String(),
+			Type:            "com.example.heartbeat",
+			Source:          sourceURI,
+			Time:            &types.Timestamp{Time: time.Now()},
+			DataContentType: strptr("application/json"),
+		}.AsV01(),
```
Comment (vaikas) [inline pkg/adapter/adapter.go]: Why is this being converted to v01 here?

File: pkg/adapter/adapter.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+var sourceURI = types.URIRef{URL: url.URL{Scheme: "http", Host: "heartbeat.example.com", Path: "/heartbeat-source"}}
+
+func strptr(s string) *string { return &s }
+
+func (a *Adapter) newEvent() cloudevents.Event {
+
+	e := cloudevents.Event{
+		Context: cloudevents.EventContextV1{
+			ID:              uuid.New().String(),
+			Type:            "com.example.heartbeat",
```
Comment (vaikas) [inline pkg/adapter/adapter.go]: should this reflect something about the samples.knative.dev here instead?

File: pkg/apis/samples/v1alpha1/samplesource_types.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	// Interval is the time interval between events.
+	//
+	// The string format is a sequence of decimal numbers, each with optional
+	// fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time
+	// units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	Interval string `json:"interval"`
+
+	// Sink is a reference to an object that will resolve to a host
+	// name to use as the sink.
+	Sink *duckv1beta1.Destination `json:"sink"`
```
Comment (vaikas) [inline pkg/apis/samples/v1alpha1/samplesource_types.go]: might as well bump to v1?

File: pkg/apis/samples/v1alpha1/samplesource_types.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+)
+
+// SampleSourceStatus communicates the observed state of the SampleSource (from the controller).
+type SampleSourceStatus struct {
+	duckv1.Status `json:",inline"`
+
+	// SinkURI is the current active sink URI that has been configured
+	// for the SampleSource.
+	// +optional
+	SinkURI string `json:"sinkUri,omitempty"`
```
Comment (vaikas) [inline pkg/apis/samples/v1alpha1/samplesource_types.go]: maybe *apis.URL instead of String?

File: config/300-samplesource.yaml [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+                      kind:
+                        type: string
+                        minLength: 1
+                      name:
+                        type: string
+                        minLength: 1
+                  uri:
+                    type: string
+                    description: "The target URI. If ref is provided, this must be a relative URI reference"
+              - type: object
```
Comment (vaikas) [inline config/300-samplesource.yaml]: maybe not support this since this is new and might be better to not start with the deprecated fields?

Comment (lberk) [reply config/300-samplesource.yaml]: ￼fixed in e7eabf7

File: pkg/apis/samples/v1alpha1/samplesource_types.go (hunk 5 above)
Comment (lberk) [reply pkg/apis/samples/v1alpha1/samplesource_types.go]: done, and propagated changes elsewhere

File: pkg/apis/samples/v1alpha1/samplesource_types.go (hunk 4 above)
Comment (lberk) [reply pkg/apis/samples/v1alpha1/samplesource_types.go]: done

File: pkg/adapter/adapter.go (hunk 3 above)
Comment (lberk) [reply pkg/adapter/adapter.go]: done in ca1ccd4

File: pkg/adapter/adapter.go (hunk 2 above)
Comment (lberk) [reply pkg/adapter/adapter.go]: Originally it was `.AsV1()` but that wasn't working with the `gcr.io/knative-releases/github.com/knative/eventing-sources/cmd/event_display` image.  I assume I should probably find an updated `event_display` image to work with cloudevent v1 spec?

**Comment by lberk** [comment]
working on the unit test failures now

File: pkg/adapter/adapter.go (hunk 2 above)
Comment (vaikas) [reply pkg/adapter/adapter.go]: @n3wscott thoughts here??

Comment (lberk) [reply pkg/adapter/adapter.go]: Aha.  When I use a non-ancient version of the event_display image, it works with `.AsV1` my mistake for using an `eventing-sources` era image.

Fixed in 5b80d6b

**Pull request merged**